File: implementations/java/ff3-validate/src/main/java/dev/horizondigital/ff3/validate/FF3Validate.java
package dev.horizondigital.ff3.validate;

import dev.horizondigital.ff3.FF3API;
import dev.horizondigital.ff3.FF3Alphabets;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.File;

@SuppressWarnings("java:S106") // System.out is appropriate for validation tool output
public class FF3Validate {
    private static final ObjectMapper objectMapper = new ObjectMapper();

    public static void main(String[] args) {
        try {
            String vectorFile = args.length > 0 ? args[0] : "../../../shared/test-vectors/nist_ff3_official_vectors.json";

            File file = new File(vectorFile);
            if (!file.exists()) {
                System.err.println("Error: Test vector file not found: " + vectorFile);
                System.exit(1);
            }

            System.out.println("FF3 NIST Validation Tool");
            System.out.println("======================");
            System.out.printf("Vector file: %s%n%n", vectorFile);

            JsonNode root = objectMapper.readTree(file);
            JsonNode vectors = root.get("vectors");
            int passed = 0;
            int total = 0;

            for (JsonNode vector : vectors) {
                total++;
                try {
                    String keyHex = vector.get("key").asText();
                    String tweakHex = vector.get("tweak").asText();
                    int radix = vector.get("radix").asInt();
                    String plaintext = vector.get("plaintext").asText();
                    String expectedCiphertext = vector.get("ciphertext").asText();

                    byte[] key = FF3API.hexToBytes(keyHex);
                    byte[] tweak = FF3API.hexToBytes(tweakHex);
                    String alphabet = getAlphabetForRadix(radix);

                    FF3API ff3 = new FF3API(key, tweak, alphabet);
                    String actualCiphertext = ff3.encrypt(plaintext);
                    String decrypted = ff3.decrypt(actualCiphertext);

                    boolean encryptPassed = expectedCiphertext.equals(actualCiphertext);
                    boolean roundtripPassed = plaintext.equals(decrypted);
                    boolean testPassed = encryptPassed && roundtripPassed;

                    if (testPassed) passed++;

                    System.out.printf("Test %2d (radix %2d): %-8s | Encrypt: %-4s | Roundtrip: %-4s%n",
                            total, radix, testPassed ? "PASS" : "FAIL",
                            encryptPassed ? "PASS" : "FAIL",
                            roundtripPassed ? "PASS" : "FAIL");

                    if (!testPassed) {
                        System.out.printf("  Input:    %s%n", plaintext);
                        System.out.printf("  Expected: %s%n", expectedCiphertext);
                        System.out.printf("  Actual:   %s%n", actualCiphertext);
                        System.out.printf("  Decrypted: %s%n", decrypted);
                        System.out.println();
                    }

                } catch (Exception e) {
                    System.out.printf("Test %2d: ERROR - %s%n", total, e.getMessage());
                }
            }

            System.out.println("======================");
            System.out.printf("Results: %d/%d passed (%.1f%%)%n", passed, total, (double) passed / total * 100);

            if (passed == total) {
                System.out.println("✓ All NIST test vectors passed!");
                System.exit(0);
            } else {
                System.out.println("✗ Some test vectors failed!");
                System.exit(1);
            }

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }

    private static String getAlphabetForRadix(int radix) {
        switch (radix) {
            case 10: return FF3Alphabets.DIGITS;
            case 16: return FF3Alphabets.HEX_LOWER;
            case 26: return FF3Alphabets.RADIX26;
            case 36: return FF3Alphabets.BASE36_LOWER;
            case 62: return FF3Alphabets.BASE62;
            default:
                throw new IllegalArgumentException("Unsupported radix: " + radix);
        }
    }
}

File: implementations/python/cli/ff3_validate.py
#!/usr/bin/env python3
"""FF3 NIST Validation Tool - Validates against official test vectors"""

import argparse
import json
import sys
import os
import binascii
from pathlib import Path

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))

from src.ff3 import (
    digits,
    base36_lower,
    base62,
    from_spec,
    AlphabetSpec,
)


def find_vectors_file(custom_path):
    """Find the test vectors file, trying multiple locations"""
    if custom_path:
        if os.path.exists(custom_path):
            return custom_path
        print(f"Error: Vectors file not found: {custom_path}", file=sys.stderr)
        sys.exit(2)

    # Try multiple paths
    paths = [
        os.environ.get('FF3_TEST_VECTORS_PATH'),
        "../../shared/test-vectors/nist_ff3_official_vectors.json",
        "../../../shared/test-vectors/nist_ff3_official_vectors.json",
        "./nist_ff3_official_vectors.json",
    ]

    for path in paths:
        if path and os.path.exists(path):
            return path

    print("Error: Could not find NIST test vectors file", file=sys.stderr)
    print("Try: ff3-validate --vectors /path/to/vectors.json", file=sys.stderr)
    sys.exit(2)


def main():
    parser = argparse.ArgumentParser(
        description="FF3 NIST Test Vector Validation Tool",
        add_help=False
    )
    parser.add_argument("--vectors", type=str, help="Path to test vectors JSON file")
    parser.add_argument("--verbose", action="store_true", help="Show detailed test output")
    parser.add_argument("--quiet", action="store_true", help="Only show failures and summary")
    parser.add_argument("-h", "--help", action="store_true", help="Show help message")

    args = parser.parse_args()

    if args.help:
        print("FF3 NIST Test Vector Validation Tool")
        print()
        print("Usage: ff3-validate [OPTIONS]")
        print()
        print("Options:")
        print("  --vectors PATH    Path to test vectors JSON file")
        print("  --verbose         Show detailed test output")
        print("  --quiet           Only show failures and summary")
        print("  -h, --help        Show this help message")
        print()
        sys.exit(0)

    if not args.quiet:
        print("FF3 NIST Validation Tool")
        print("=" * 40)

    vectors_path = find_vectors_file(args.vectors)

    if not args.quiet:
        print(f"Vector file: {vectors_path}")
        print()

    try:
        with open(vectors_path, 'r') as f:
            data = json.load(f)
            vectors = data.get("vectors", [])
    except FileNotFoundError:
        print(f"Error: Test vector file not found: {vectors_path}", file=sys.stderr)
        sys.exit(2)
    except json.JSONDecodeError as e:
        print(f"Error: Invalid JSON in test vector file: {e}", file=sys.stderr)
        sys.exit(1)

    if not args.quiet:
        print(f"Testing {len(vectors)} NIST FF3 vectors...")
        print()

    passed = 0
    total = 0

    for vector in vectors:
        total += 1
        try:
            key = binascii.unhexlify(vector["key"])
            tweak = binascii.unhexlify(vector["tweak"])
            radix = vector["radix"]
            plaintext = vector["plaintext"]
            expected = vector["ciphertext"]

            # Get appropriate cipher based on radix
            if radix == 10:
                cipher = digits(key, tweak)
            elif radix == 26:
                # Use a custom radix-26 alphabet (matching other implementations)
                spec = AlphabetSpec("0123456789abcdefghijklmnop")
                cipher = from_spec(key, tweak, spec)
            elif radix == 36:
                cipher = base36_lower(key, tweak)
            elif radix == 62:
                cipher = base62(key, tweak)
            else:
                raise ValueError(f"Unsupported radix: {radix}")

            # Test encryption
            actual = cipher.encrypt(plaintext)
            encrypt_passed = actual == expected

            # Test roundtrip
            decrypted = cipher.decrypt(actual)
            roundtrip_passed = decrypted == plaintext

            test_passed = encrypt_passed and roundtrip_passed

            if test_passed:
                passed += 1
                if args.verbose:
                    print(f"Sample {vector.get('sample', total)}: PASS")
            else:
                if not args.quiet:
                    print(f"Sample {vector.get('sample', total)}: FAIL")
                    if not encrypt_passed:
                        print(f"  Expected:  {expected}")
                        print(f"  Got:       {actual}")
                    if not roundtrip_passed:
                        print(f"  Round-trip failed")
                    print()

        except Exception as e:
            if not args.quiet:
                print(f"Sample {vector.get('sample', total)}: ERROR - {e}")

    if not args.quiet:
        print("=" * 40)
        print(f"Results: {passed}/{total} passed ({passed/total*100:.1f}%)")
        print()

    if passed == total:
        if not args.quiet:
            print("ALL NIST TEST VECTORS PASSED!")
            print()
            print("WARNING: FF3 was WITHDRAWN by NIST due to security vulnerabilities.")
            print("This implementation is for EDUCATIONAL and RESEARCH purposes only.")
            print("DO NOT use in production systems.")
            print()
        sys.exit(0)
    else:
        if not args.quiet:
            print("VALIDATION FAILED")
        sys.exit(1)


if __name__ == "__main__":
    main()
